mlir/lib/Dialect/Linalg/Transforms/WinogradConv2D.cpp

```diff
@@ -36,6 +190,329 @@ constexpr TransformMapKeyTy F_2_3{2, 3};
+  static const llvm::SmallDenseMap<TransformMapKeyTy, TransformMatrix>
+      BMatrices = {
+          {F_2_3, TransformMatrix(B_2x2_3x3, 4, 4)},
+          {F_4_3, TransformMatrix(B_4x4_3x3, 6, 6)},
+          {F_2_5, TransformMatrix(B_2x2_5x5, 6, 6)},
+      };
+
+  auto inputType = cast<ShapedType>(input.getType());
+  Type elementType = inputType.getElementType();
+  auto inputShape = inputType.getShape(); // N, H, W, C
+  int64_t inputN = inputShape[0];
+  int64_t inputH = inputShape[1];
+  int64_t inputW = inputShape[2];
+  int64_t inputC = inputShape[3];
+  int64_t alphaH = leftTransform ? m + r - 1 : 1;
+  int64_t alphaW = rightTransform ? m + r - 1 : 1;
+
+  if (inputH != alphaH && inputH != 1)
+    return Value();
+  if (inputW != alphaW && inputW != 1)
+    return Value();
+
+  auto buildBody = [&](OpBuilder &builder, Location loc, ValueRange ivs,
+                       ValueRange args) -> scf::ValueVector {
+    Value NIter = ivs[0];
+    Value CIter = ivs[1];
+
+    // Extract (H, W) from (N, H, W, C).
+    auto extractInput = extract2DData(
+        builder, loc, input, NIter, CIter, /*outLoopIdx=*/0,
+        /*inLoopIdx=*/3, /*heightIdx=*/1, /*widthIdx=*/2, /*srcSize=*/4);
+
+    TransformMapKeyTy key = {m, r};
+    int64_t retRows = 1;
+    int64_t retCols = 1;
+    Value matmulRetValue = extractInput;
+    if (leftTransform) {
+      // Get constant transform matrix BT.
+      auto it = BTMatrices.find(key);
+      if (it == BTMatrices.end())
+        return {};
+      const TransformMatrix &BTMatrix = it->second;
+
+      retRows = BTMatrix.rows;
+      auto matmulType = RankedTensorType::get({retRows, inputW}, elementType);
+      auto init = builder.create<tensor::EmptyOp>(loc, matmulType.getShape(),
+                                                  elementType);
+
+      Value BT =
+          create2DTransformMatrix(builder, loc, BTMatrix, builder.getF32Type());
+      // Multiply BT x d.
+      auto matmulOp = builder.create<linalg::MatmulOp>(
+          loc, matmulType, ValueRange{BT, matmulRetValue}, ValueRange{init});
+      matmulRetValue = matmulOp.getResult(0);
+    }
+
+    if (rightTransform) {
+      // Get constant transform matrix B.
+      auto it = BMatrices.find(key);
+      if (it == BMatrices.end())
+        return {};
+      const TransformMatrix &BMatrix = it->second;
+
+      retCols = BMatrix.cols;
+      auto matmulType = RankedTensorType::get({retRows, retCols}, elementType);
+      auto init = builder.create<tensor::EmptyOp>(loc, matmulType.getShape(),
+                                                  elementType);
+      Value B =
+          create2DTransformMatrix(builder, loc, BMatrix, builder.getF32Type());
+      // Multiply v = (BT x d) x B.
+      auto matmulOp = builder.create<linalg::MatmulOp>(
+          loc, matmulType, ValueRange{matmulRetValue, B}, ValueRange{init});
+      matmulRetValue = matmulOp.getResult(0);
+    }
+
+    // Insert (H, W) to (H, W, 1, 1, N, C).
```

Comment:
The shape of the destination here should not always be `1x1` in the inner H and W dims. There are 2 ways that I would suggest for implementing the decomposition here:

1. Before decomposing, tile `N`, `C`, `HDiv`, `WDiv` down to `1`, and require these sizes to be `1` when you do the decomposition. The decomposition is very simple then, because it does not need any loop nest. This is what we have done in IREE, and the requirement to tile these dims to 1 before decomposing has not posed any problems so far.
2. The other way to do it would be to allow for partial tiling of `N`, `C`, `HDiv`, `WDiv`, and do the remaining tiling down to 1 as part of the decomposition. This approach makes the tile size selection before decomposition a little more flexible, but you need to be careful that the tiling implementation tiles `HDiv` and `WDiv` of the result instead of `H` and `W` of the input (this is natural since tiling implementations usually tile based on the result, but in the output transform things are reversed, and you would want to tile based on the input).

What is being done by this PR seems to be sort of a mix of the 2. The `HDiv` and `WDiv` dimensions are required to be 1, but the `N` and `C` dimensions are not. If you are going to restrict dimensions, then restrict them all to be `1` before decomposition, and just make the decomposition simpler, since the tile size selection will need to be aware of this requirement regardless of how many dimensions are required to be `1`. Option (2) is the more general way to do it, but it makes the decomposition a bit more complex (although you already have implemented some of the complexity here, so maybe worth trying). I will leave the decision between (1) and (2) up to you, but I would recommend picking one or the other.

Reply:
The implementation is based on the pseudo code of its origin paper which describes how to support arbitrary size of height and width using tiling and using nested loops for N and C. That's why my implementation only considers to tile height and width outside decomposition. Your proposal 2 is interesting that to support arbitrary values of tileH and tileW inside decomposition. I can give it a try to make the decomposition more flexible.

Thanks for your review and suggestion.

Reply:
After trying to deal with tileH, and tileW in (alphH, alphaW, tileH, tileW, N, C) inside input decomposition, we need to deal with tiling for H and W when iterating through tileH and tileW. However, there is similar logic implemented in TilingInterface already. I suggest to keep decomposition simple and avoid the duplication for tiling H and W in decomposition. For N and C, there is no relation to H and W. So, it is simple to keep it inside decomposition and make it flexible as current implementation. I suggest to keep my original implementation. What do you think, @Max191?

Reply:
I would strongly recommend supporting tiling of all dimensions (except got alphaH, alphaW) in the TilingInterface implementation. TilingInterface is very useful for reducing problem sizes to shapes that fit nicely into caches and vector registers. Having control over all dimensions in the TilingInterface is very important for being able to do this. In the current implementation, the tiles will always contain the full size N and C dimensions (until the inner tile after decomposition, which will have a size of 1), which greatly restricts the options for tile size selection.

I would either refactor the logic into a shared utility function, or move the logic into the TilingInterface implementation. You can add the util function here if you decide to do that:
https://github.com/llvm/llvm-project/blob/main/mlir/lib/Dialect/Linalg/Utils/Utils.cpp

Reply:
I updated the implementation to tile height and width inside decomposition.